src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -1517,7 +1534,34 @@ namespace winrt::TerminalApp::implementation
                     text = item.Path();
                 }
             }
-            eventArgs.HandleClipboardData(text);
+
+            co_await winrt::resume_foreground(Dispatcher());
+
+            const bool hasNewLine = std::find(text.cbegin(), text.cend(), L'\n') != text.cend();
+            const bool warnMultiLine = hasNewLine && _settings->GlobalSettings().WarnAboutMultiLinePaste();
+
+            constexpr const std::size_t minimumSizeForWarning = 1024 * 5; // 5 KiB
+            const bool warnLargeText = text.size() > minimumSizeForWarning &&
+                                       _settings->GlobalSettings().WarnAboutLargePaste();
+
+            if (!warnMultiLine && !warnLargeText)
+            {
+                eventArgs.HandleClipboardData(text);
+                co_return;
+            }
+
+            _acceptPaste = [=]() {
```

Comment:
I'm wondering if we need an overarching design change that lets us pass a completion lambda into the dialog _or perhaps_ we should make ShowDialog return an IAsyncOperation that we can co_await..

I don't love stashing state outside our flow of control for this

Reply:
Ok. I've modified a little bit `AppLogic` so that it implements a new interface called `IDialogPresenter` with a method `ShowDialog` which returns an `IAsyncOperation`, and now `TerminalPage` stores a `IDialogPresenter`.